File: backend/src/main/java/com/grepp/smartwatcha/app/model/admin/movie/upcoming/service/common/UpcomingMovieFetchService.java
package com.grepp.smartwatcha.app.model.admin.movie.upcoming.service.common;

import com.grepp.smartwatcha.app.controller.api.admin.upcoming.api.UpcomingMovieApi;
import com.grepp.smartwatcha.app.controller.api.admin.upcoming.payload.UpcomingMovieApiResponse;
import com.grepp.smartwatcha.app.model.admin.movie.upcoming.dto.UpcomingMovieDto;
import com.grepp.smartwatcha.app.controller.api.admin.upcoming.payload.UpcomingMovieCreditApiResponse;
import com.grepp.smartwatcha.app.controller.api.admin.upcoming.payload.UpcomingMovieDetailApiResponse;
import com.grepp.smartwatcha.app.controller.api.admin.upcoming.payload.UpcomingMovieReleaseDateApiResponse;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

/*
 * 공개 예정작 영화 정보 조회 서비스
 * TMDB API 를 통해 영화 정보를 조회하고, 병렬로 상세 정보를 가져와 DTO 를 구성
 *
 * 주요 기능:
 * - 영화 기본 정보 조회
 * - 병렬로 크레딧, 개봉일, 상세 정보 조회
 * - 조회된 정보를 DTO 에 통합
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class UpcomingMovieFetchService { // "영화 1편"의 정보를 한 DTO 에 채워주는 흐름 관리용 서비스

  private final UpcomingMovieAsyncFetchService asyncService;
  private final UpcomingMovieDetailEnricher dtoEnricher;
  private final UpcomingMovieApi upcomingMovieApi;

  @Value("${tmdb.api.key}")
  private String apiKey;

  // 영화의 기본 정보를 바탕으로 상세 정보를 병렬로 조회하여 DTO 를 구성
  public UpcomingMovieDto buildEnrichedDto(UpcomingMovieDto baseDto, String apiKey) throws ExecutionException, InterruptedException {
    Long movieId = baseDto.getId();

    // 병렬 호출 시작
    CompletableFuture<UpcomingMovieCreditApiResponse> creditsFuture = asyncService.fetchCredits(movieId, apiKey);
    CompletableFuture<UpcomingMovieReleaseDateApiResponse> releaseFuture = asyncService.fetchReleaseDates(movieId, apiKey);
    CompletableFuture<UpcomingMovieDetailApiResponse> detailFuture = asyncService.fetchDetails(movieId, apiKey);

    // 모든 호출 완료될 때까지 대기
    CompletableFuture.allOf(creditsFuture,releaseFuture, detailFuture).join();

    // 결과 꺼내기 - 실패한 항목은 null 처리
    UpcomingMovieCreditApiResponse credits = safeGet(creditsFuture, "credits", movieId);
    UpcomingMovieReleaseDateApiResponse releaseDate = safeGet(releaseFuture, "releaseDates", movieId);
    UpcomingMovieDetailApiResponse details = safeGet(detailFuture, "details", movieId);

    // enrich 단계 - null 허용 (일부 enrich 실패해도 계속 진행)
    if(credits != null) {
      dtoEnricher.enrichCredits(baseDto, credits);
    }
    if(releaseDate != null) {
      dtoEnricher.enrichCertification(baseDto, releaseDate);
    }
    if(details != null) {
      dtoEnricher.enrichCountry(baseDto, details);
    }
    return baseDto;
  }

  // 안전하게 CompletableFuture 의 결과를 가져오는 유틸리티 메서드
  private <T> T safeGet(CompletableFuture<T> future, String apiName, Long movieId) {
    try {
      return future.get();
    } catch (InterruptedException | ExecutionException e) {
      log.warn("⚠️ [{}] API 실패 - movieId: {} → enrich 생략됨", apiName, movieId);
      return null;
    }
  }

  // TMDB API 를 통해 공개 예정작 영화 목록을 페이지네이션하여 조회
  public List<UpcomingMovieDto> fetchUpcomingMovies() {
    int page = 1;
    UpcomingMovieApiResponse response = upcomingMovieApi.getUpcomingMovies(apiKey, "en-US", page, "US");
    int totalPages = response.getTotalPages();

    List<UpcomingMovieDto> allMovies = new java.util.ArrayList<>(response.getMovies());
    for (page = 2; page <= totalPages; page++) {
      UpcomingMovieApiResponse nextPage = upcomingMovieApi.getUpcomingMovies(apiKey, "en-US", page, "US");
      allMovies.addAll(nextPage.getMovies());
    }
    return allMovies;
  }
}

File: userserviceserver/src/main/kotlin/com/grepp/smartwatcha/app/model/upcomingmovie/service/UpcomingMovieFetchService.kt
package com.grepp.smartwatcha.app.model.upcomingmovie.service

import com.grepp.smartwatcha.app.controller.api.upcomingmovie.api.*
import com.grepp.smartwatcha.app.controller.api.upcomingmovie.payload.*
import com.grepp.smartwatcha.app.model.upcomingmovie.dto.UpcomingMovieDto
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.withContext
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import java.time.LocalDate
import java.time.format.DateTimeFormatter

// TMDB 공개 예정작 영화 조회 및 상세 정보 병합 서비스
// - TMDB API로부터 공개 예정작 목록을 조회하고, 각 영화에 대한 상세 정보를 병렬로 조회하여 DTO에 보강한다.
// - 주요 보강 항목: 크레딧(배우/감독/작가), 등급/개봉유형, 제작국가
@Service
class UpcomingMovieFetchService(
    private val upcomingMovieApi: UpcomingMovieApi,
    private val creditApi: UpcomingMovieCreditApi,
    private val releaseDateApi: UpcomingMovieReleaseDateApi,
    private val detailApi: UpcomingMovieDetailApi,
    private val dtoEnricher: UpcomingMovieDetailEnricher
) {
    @Value("\${tmdb.api.key}")
    private lateinit var apiKey: String

    private val log = LoggerFactory.getLogger(UpcomingMovieFetchService::class.java)

    // TMDB로부터 전체 공개 예정작 목록을 페이징 방식으로 조회
    // - 날짜가 오늘 이후인 영화만 필터링하여 반환
    // - 각 페이지를 순차적으로 호출하여 누적
    fun fetchUpcomingMovies(): List<UpcomingMovieDto> {
        val page = 1
        val response: UpcomingMovieApiResponse =
            upcomingMovieApi.getUpcomingMovies(apiKey, "en-US", page, "US")
        val totalPages = response.totalPages

        val allMovies = response.movies.toMutableList()
        for (p in 2..totalPages) {
            val nextPage = upcomingMovieApi.getUpcomingMovies(apiKey, "en-US", p, "US")
            allMovies.addAll(nextPage.movies)
        }

        // 오늘 이후 개봉하는 영화만 필터링
        val today = LocalDate.now()
        val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd")

        val upcomingOnly = allMovies.filter { dto ->
            try {
                val parsedDate = LocalDate.parse(dto.releaseDate, formatter)
                parsedDate.isAfter(today)
            } catch (e: Exception) {
                false // 잘못된 날짜 형식은 필터링 제외
            }
        }

        log.info("🎯 TMDB에서 받은 영화 ${allMovies.size}건 중, 개봉 예정 ${upcomingOnly.size}건 필터링")

        return upcomingOnly
    }

    // 단일 영화 DTO에 대한 상세 정보 보강 (크레딧, 등급, 국가)
    // - 코루틴 기반으로 병렬 fetch → DTO 보강 수행
    suspend fun buildEnrichedDto(baseDto: UpcomingMovieDto, apiKey: String): UpcomingMovieDto? {
        val movieId = baseDto.id

        return withContext(Dispatchers.IO) {
            val (credits, releaseDates, details) = fetchAll(movieId, apiKey)

            credits?.let { dtoEnricher.enrichCredits(baseDto, it) }
            releaseDates?.let { dtoEnricher.enrichCertification(baseDto, it) }
            details?.let { dtoEnricher.enrichCountry(baseDto, it) }

            baseDto
        }
    }

    // 단일 영화에 대한 3종 API 병렬 호출 (크레딧, 릴리즈, 디테일)
    // - 예외 발생 시 로그만 남기고 null 처리
    private suspend fun fetchAll(
        movieId: Long,
        apiKey: String
    ): Triple<UpcomingMovieCreditApiResponse?, UpcomingMovieReleaseDateApiResponse?, UpcomingMovieDetailApiResponse?> =
        coroutineScope {
            val creditDeferred = async {
                try {
                    creditApi.getCredits(movieId, apiKey)
                } catch (e: Exception) {
                    log.error("❌ [fetchCredits] 실패 - movieId: {}", movieId, e)
                    null
                }
            }

            val releaseDeferred = async {
                try {
                    releaseDateApi.getReleaseDates(movieId, apiKey)
                } catch (e: Exception) {
                    log.error("❌ [fetchReleaseDates] 실패 - movieId: {}", movieId, e)
                    null
                }
            }

            val detailDeferred = async {
                try {
                    detailApi.getDetails(movieId, apiKey)
                } catch (e: Exception) {
                    log.error("❌ [fetchDetails] 실패 - movieId: {}", movieId, e)
                    null
                }
            }

            Triple(creditDeferred.await(), releaseDeferred.await(), detailDeferred.await())
        }
}
